dynawo/sources/Models/CPP/ModelLoadRestorativeWithLimits/DYNModelLoadRestorativeWithLimits.cpp

```diff
@@ -0,0 +1,577 @@
+      jt.addTerm(UrYNum_ + rowOffset, - Q_dUr(Ur(), Ui(), U(), U2()));
+      jt.addTerm(UiYNum_ + rowOffset, - Q_dUi(Ur(), Ui(), U(), U2()));
+      jt.addTerm(UfYNum_ + rowOffset, - Q_dUf(U()));
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalG(const double& /*t*/) {
+    gLocal_[0] = (doubleNotEquals(UfRaw(), UMaxPu_) && UfRaw() > UMaxPu_) ? ROOT_UP : ROOT_DOWN;
+    gLocal_[1] = (doubleNotEquals(UfRaw(), UMinPu_) && UMinPu_ - UfRaw() > 0) ? ROOT_UP : ROOT_DOWN;
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::setGequations() {
+    if (getIsInitProcess())
+      return;
+    if (isConnected()) {
+      if (gLocal_[0] == ROOT_UP) {
+        gEquationIndex_[0] = std::string("UMaxPu limit reached by Load: ").append(staticId());
+      }
+      if (gLocal_[1] == ROOT_UP) {
+        gEquationIndex_[1] = std::string("UMinPu limit reached by Load: ").append(staticId());
+      }
+    }
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::evalJtPrim(const double & /*t*/, const double & /*cj*/, SparseMatrix& jt, const int& rowOffset) {
+    jt.changeCol();
+    if (isConnected())
+      jt.addTerm(UfRawYNum_ + rowOffset, Tf_);
+    else
+      jt.addTerm(UfRawYNum_ + rowOffset, 1);
+    jt.changeCol();
+    jt.changeCol();
+    jt.changeCol();
+    jt.changeCol();
+    jt.changeCol();
+  }
+
+  modeChangeType_t
+  ModelLoadRestorativeWithLimits::evalMode(const double& /*t*/) {
+    if (gLocal_[0] == ROOT_UP || gLocal_[1] == ROOT_UP)
+      return ALGEBRAIC_MODE;
+    if (connectionState_prev != getConnected()) {
+      connectionState_prev = getConnected();
+      return ALGEBRAIC_J_UPDATE_MODE;
+    }
+    return NO_MODE;
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::getY0() {
+    unsigned int index = 0;
+    // ufRaw
+    yLocal_[index] = u0_;
+    ypLocal_[index] = UfRawprim0_;
+    // uf
+    ++index;
+    yLocal_[index] = u0_;
+    ypLocal_[index] = 0.;
+    // ur
+    ++index;
+    // ui
+    ++index;
+    // Ir
+    ++index;
+    yLocal_[index] = (P0_ * u0_ * cos(angleO_) + Q0_ * u0_ * sin(angleO_)) / (u0_ * u0_);
+    ypLocal_[index] = 0;
+    // Ii
+    ++index;
+    yLocal_[index] = (P0_ * u0_ * sin(angleO_) - Q0_ * u0_ * cos(angleO_)) / (u0_ * u0_);
+    ypLocal_[index] = 0;
+    // P;
+    ++index;
+    yLocal_[index] = P0_;
+    ypLocal_[index] = 0.;
+    // Q;
+    ++index;
+    yLocal_[index] = Q0_;
+    ypLocal_[index] = 0.;
+
+    zLocal_[0] = false;
+    zLocal_[1] = false;
+  }
+
+  void
+  ModelLoadRestorativeWithLimits::initializeFromData(const boost::shared_ptr<DataInterface>& /*data*/) {
+    /* not need */
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::P_dUr(const double& ur, const double& /*ui*/, const double& U, const double& U2) const {
+    return P0_ * alpha_ * ur * pow(U/Uf(), alpha_) / U2;
+  }
+
+  double
+  ModelLoadRestorativeWithLimits::P_dUi(const double& /*ur*/, const double& ui, const double& U, const double& U2) const {
+    return P0_ * alpha_ * ui * pow(U/Uf(), alpha_) / U2;
+  }
```

Comment:
You certainly call them from the same method (evalJt or evalJTPrim): what about calculating the pow(U/Uf(), alpha_) once at the upper level?

Reply:
From a performance perspective I don't think declaring multiple functions to compute parts of evalF or evalJ or retrieve yLocal_[index] is a good practice, better do it once ine the dedicated method (evalF or evalJ) an factorize as much as possible heavy operations like pow. From a "readable" perspective mayber it is good to have small functions